Gems you want to have in your Rails API (#87)
rootstrap/blog: content/Gems You Want To Have In Your Rails API - Juan Diego Rosales.md

# Gems You Want To Have In Your Rails API

> *Juan Diego Rosales is a software developer on the Rootstrap team and writes about his experience working on different projects. You can also follow Juan on [LinkedIn](www.linkedin.com/in/juan-diego-rosales-02a05a144).*

Reviewer: shouldn't this be in the `About` section, for example here
![image](https://user-images.githubusercontent.com/6373536/78814037-6556b480-79a4-11ea-92c9-dad4cc5d30fb.png)

Are you new to Rails? Or maybe just looking for any cool gem to use in your new API? Here is a list of gems we have on our Rails API Base that we consider a must-have in every Rails API project.

## Authentication

> If you are building a Rails API, then you will probably need token based authentication and that is when [Devise Token Auth](https://github.com/lynndylanhurley/devise_token_auth) comes in. This gem is built on top of Devise, one of the most popular authentication gems for Rails which you may have already worked with. If that is the case, then you'll get familiar with Devise Token Auth very easily.

Reviewer 2: ```suggestion
If you are building a Rails API, then you will probably need token-based authentication and that is when [Devise Token Auth](https://github.com/lynndylanhurley/devise_token_auth) comes in. This gem is built on top of Devise, one of the most popular authentication gems for Rails which you may have already worked with. If that is the case, then you'll get familiar with Devise Token Auth very easily.
```

With very little setup, this gem provides registration, sign in, sign out and reset password flows out of the box, as well as very useful helper functions. Also, it handles email confirmation if enabled and supports the use of multiple user models, giving you the possibility to manage the authentication of different types of users at the same time using groups.

> In comparison to other alternatives available, one of the best perks that comes with Devise Token Auth, which is inherited from Devise, is the very little effort you have to invest to get things up and running. On top of that, you can also customize the different flows mentioned if you want to offer a different experience, and while this may be a little more challenging, it's not rocket science!

Reviewer: I think it's already clear that it's `inherited from Devise` 😛

## Resource administration

> Most APIs typically need some level of resource administration and for that I would definitely recommend [Active Admin](https://github.com/activeadmin/activeadmin). Despite the fact it can have a steep learning curve, specially at the beginning, due to it's DSL (domain-specific language), once you get used to it you can have admin pages with CRUD actions for any resource in no time.

Reviewer 2: Suggestion: for someone who isn't familiar with the DSL acronym, it might be of help pointing to a summary of the concept rather than just unfurl it. It doesn't have to be the MF article, it could also be Wikipedia or any other resource that provides the needed clarification.

```suggestion
Most APIs typically need some level of resource administration and for that I would definitely recommend [Active Admin](https://github.com/activeadmin/activeadmin). Despite the fact it can have a steep learning curve, specially at the beginning, due to it's [DSL](https://martinfowler.com/dsl.html), once you get used to it you can have admin pages with CRUD actions for any resource in no time.
```

Author: Good suggestion! 👍

> Something very important to mention is that Active Admin has a large community and good documentation relative to other administration gems, such as Rails Admin or Administrate. Also, another strong feature is that you can configure the index filters according to your needs and they can be customized to use more that one attribute since they support [Ransack](https://github.com/activerecord-hackery/ransack). 

Reviewer: `relative to` ? or `compared to` ?

Author: `compared to` is more suitable, nice!

If you need a custom page, you can have Active Admin render your own partial and you will probably need to make changes in the controller actions. This is not an easy task, but none of the other gems offer anything better.

> For some people, one of the downsides could be the design, since Active Admin does not come with a powerful CSS framework out of the box. Luckily, there are many complementary gems that provide a better looking theme, for example [Arctic Admin](https://github.com/cprodhomme/arctic_admin).

Reviewer 3: what about referencing all the options besides giving a specific example: https://github.com/activeadmin/activeadmin/wiki/Themes ?

## Testing
Although it may take some time to get used to the DSL, here at Rootstrap we choose [RSpec](https://github.com/rspec/rspec-rails) for testing. The main reason for this is that it encourages human readable tests, which ends up helping a lot in the development process.

When writing tests you always need to set up database records to create the context, and for this we like to use factories with [Factory Bot](https://github.com/thoughtbot/factory_bot). This gem offers great flexibility, allowing the developer to easily create different scenarios, but also improves the readability as you can easily know what is going on in the test.

Last but not least, if your API relies on third party services you will need stubbing on the external HTTP requests and for this I recommend using [Webmock](https://github.com/bblimke/webmock). If you ask me, I found this gem much simpler and flexible than VCR, for example.

## Linting

> Finally, every developer wants a good quality codebase and in that sense linters are essential. Here's a list of gems to consider:

Reviewer: ```suggestion
Finally, every developer wants a good quality codebase and in that sense linters are essential. Here's a list of gems that I strongly recommend:
```

* [RuboCop](https://github.com/rubocop-hq/rubocop-rails)
* [Rails Best Practices](https://github.com/flyerhzm/rails_best_practices)
* [Reek](https://github.com/troessner/reek) for code smells
* [Bullet](https://github.com/flyerhzm/bullet) for query optimization

> * [i18n-tasks](https://github.com/glebm/i18n-tasks) for translations linting

Reviewer 3: whats the difference with our i18n_linter ?

Author: I recommended the one that was in our Rails API base, and I have worked with it as well. I didn't know about that gem of ours...

Reviewer 3: As we discussed go ahead with your proposed gem 👍

With very little setup, this gem provides registration, sign in, sign out and reset password flows out of the box, as well as very useful helper functions. Also, it handles email confirmation if enabled and supports the use of multiple user models, giving you the possibility to manage the authentication of different types of users at the same time using groups.

In comparison to other alternatives available, one of the best perks that comes with Devise Token Auth is the very little effort you have to invest to get things up and running. On top of that, you can also customize the different flows mentioned if you want to offer a different experience, and while this may be a little more challenging, it's not rocket science!

## Authorization
I guess authorizing every request on an API can be achieved in the controller, but that is not scalable at all and may ease the introduction of vulnerabilities in an application. Here at Rootstrap, we find [Pundit](https://github.com/varvet/pundit) to be the perfect ally against those issues.

> Not only you will have one place in the codebase for the authorization logic, making the system more robust, but also it will help to clean up the controllers, improving the readability.

Reviewer 4: I'm not 100% sure but I think it should be `Not only will you have`

Reviewer 5: Agree
